implementations/rust/ockam/ockam_api/src/kafka/secure_channel_map.rs

```diff
@@ -0,0 +1,387 @@
+use ockam_node::compat::tokio::sync::Mutex;
+use ockam_node::Context;
+use serde::{Deserialize, Serialize};
+
+///Offer simple APIs to encrypt and decrypt kafka messages.
+/// Underneath it create secure channels for each topic/partition
+/// and uses them to encrypt the content.
+/// Multiple secure channels may be created for the same topic/partition
+/// but each will be explicitly label.
+/// It's the same for both producer and consumer although it could be split
+/// into two distinct implementation.
+///This is proxy trait to avoid propagating the vault implementation.
+#[async_trait]
+pub(crate) trait KafkaSecureChannelController: Send + Sync {
+    ///encrypt the content specifically for the consumer waiting for that topic name and partition
+    /// the the secure channel is missing it'll be created, this may take up to few seconds
+    /// the first time
+    async fn encrypt_content_for(
+        &self,
+        context: &mut Context,
+        topic_name: &str,
+        partition_id: i32,
+        content: Vec<u8>,
+    ) -> Result<(UniqueSecureChannelId, Vec<u8>), Error>;
+
+    ///decrypt the content based on the unique secure channel identifier
+    /// the secure channel is expected to be already initialized
+    async fn decrypt_content_for(
+        &self,
+        context: &mut Context,
+        secure_channel_id: UniqueSecureChannelId,
+        encrypted_content: Vec<u8>,
+    ) -> Result<Vec<u8>, Error>;
+
+    ///start forwarders in the orchestrator for each {partition}_{topic_name} combination
+    /// should be used only by the consumer.
+    /// does nothing if they were already created, but fails it they already exist.
+    async fn start_forwarders_for(
+        &self,
+        context: &mut Context,
+        topic_id: &str,
+        partitions: Vec<i32>,
+    ) -> Result<(), Error>;
+}
+
+pub(crate) struct KafkaSecureChannelControllerImpl<V: IdentityVault, S: AuthenticatedStorage> {
+    inner: Arc<Mutex<InnerSecureChannelControllerImpl<V, S>>>,
+}
+
+//had to manually implement since #[derive(Clone)] doesn't work well in this situation
+impl<V: IdentityVault, S: AuthenticatedStorage> Clone for KafkaSecureChannelControllerImpl<V, S> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+///Unique identifier for a specific secure_channel.
+/// Used in order to distinguish between secure channels created between
+/// the same identities.
+#[derive(Debug, Clone, Serialize, Deserialize, Message)]
+struct SecureChannelIdentifierMessage {
+    secure_channel_identifier: UniqueSecureChannelId,
+}
+
+pub(crate) type UniqueSecureChannelId = u64;
+type TopicPartition = (String, i32);
+struct InnerSecureChannelControllerImpl<V: IdentityVault, S: AuthenticatedStorage> {
+    //we are using encryptor api address as unique _local_ identifier
+    //of the secure channel
+    id_encryptor_map: HashMap<UniqueSecureChannelId, Address>,
```

Comment:
every map has different key,value

Reply:
My two cents: not a huge fan of maps - in terms of memory&time it's often worse (and causes memory segmentation) when there are not many elements. But sometimes it makes code simpler, then it's fine

Reply:
> every map has different key,value

@davide-baldo just to make sure I understood your answer. My comment was saying "if there is only one UniqueSecureChannelId for each Address in the id_encryptor_map, then we could enforce this invariant with a bimap instead of a hashmap". Are you saying that this is not the case?

Reply:
We could enforce it by aligning keys and values, but we would need to send extra information just for the sake of it.
The 2 maps are actually completely independent (because I used only 1 implementation for both consumer and producer to keep it simple, maybe in the future I'll split them)

Reply:
I still don't understand your answer 😬, so I will try one last time, this is not a major thing after all: is there a bijection between `UniqueSecureChannelId` and `Address` in the `id_encryptor_map`?

Reply:
Yes, there is a bijection between `UniqueSecureChannelId` and `Address`.

For each secure channel id that is created, there is a relative encryptor.

Reply:
ps: I'll try to split the file into consumer and producer structures, after the refactor it'll be clearer.

Reply:
> For each secure channel id that is created, there is a relative encryptor.

Ok, thanks. This is why I was suggesting to use a Bimap in that case instead of a HashMap :-). Not a big deal but I found it useful in the past to make this kind of relationship more precise.

Reply:
Won't it index both key and value? I only need to retrieve data one way